File: src/tui/connection_status.py
"""Format connection status lines for the TUI."""

from __future__ import annotations

from typing import Any


def _flag(connected: bool) -> str:
    return "connected" if connected else "DISCONNECTED"


def format_connection_status(info: dict[str, Any]) -> str:
    lines: list[str] = []

    if info.get("gateway_connected"):
        lines.append(f"Gateway: {_flag(True)}  {info['gateway_url']}")
    else:
        err = info.get("gateway_error") or "unreachable"
        lines.append(f"Gateway: {_flag(False)}  {info['gateway_url']}  ({err})")

    cp_url = info.get("control_plane_url")
    if not info.get("gateway_connected"):
        lines.append("Control plane: (unknown — gateway unreachable)")
    elif cp_url:
        if info.get("control_plane_connected"):
            lines.append(f"Control plane: {_flag(True)}  {cp_url}")
        else:
            err = info.get("control_plane_error") or "unreachable"
            lines.append(f"Control plane: {_flag(False)}  {cp_url}  ({err})")
    else:
        lines.append("Control plane: (not reported by gateway)")

    node_id = info.get("node_id")
    if node_id and info.get("gateway_connected"):
        running = info.get("running_agent_count")
        total = info.get("agent_count")
        if running is not None and total is not None:
            lines.append(f"Node: {node_id}  ·  {running}/{total} agents running")

    return "\n".join(lines)


def format_connection_subtitle(info: dict[str, Any]) -> str:
    if not info.get("gateway_connected"):
        return f"Gateway DISCONNECTED · {info['gateway_url']}"

    cp_url = info.get("control_plane_url") or "?"
    if info.get("control_plane_connected"):
        return f"Gateway OK · Control plane OK · {cp_url}"

    return f"Gateway OK · Control plane DISCONNECTED · {cp_url}"

File: web/src/connectionStatus.ts
export type ConnectivityInfo = {
  gateway_url?: string;
  gateway_connected: boolean;
  gateway_error?: string | null;
  control_plane_url?: string | null;
  control_plane_connected?: boolean;
  control_plane_error?: string | null;
  node_id?: string | null;
  agent_count?: number | null;
  running_agent_count?: number | null;
};

export function formatConnectionStatus(info: ConnectivityInfo): string[] {
  const lines: string[] = [];

  if (info.gateway_connected) {
    lines.push(`Gateway: ✅ connected · ${info.gateway_url ?? "?"}`);
  } else {
    const err = info.gateway_error || "unreachable";
    lines.push(`Gateway: 🔴 DISCONNECTED · ${info.gateway_url ?? "?"} (${err})`);
  }

  if (!info.gateway_connected) {
    lines.push("Control plane: (unknown — gateway unreachable)");
  } else if (info.control_plane_url) {
    if (info.control_plane_connected) {
      lines.push(`Control plane: ✅ connected · ${info.control_plane_url}`);
    } else {
      const err = info.control_plane_error || "unreachable";
      lines.push(`Control plane: 🔴 DISCONNECTED · ${info.control_plane_url} (${err})`);
    }
  } else {
    lines.push("Control plane: (not reported by gateway)");
  }

  if (info.node_id && info.gateway_connected) {
    const running = info.running_agent_count ?? "?";
    const total = info.agent_count ?? "?";
    lines.push(`Node: ${info.node_id} · ${running}/${total} agents running`);
  }

  return lines;
}

export async function fetchConnectivity(): Promise<ConnectivityInfo> {
  try {
    const status = await fetch("/api/status").then((r) => {
      if (!r.ok) throw new Error(r.statusText);
      return r.json();
    });
    return {
      gateway_url: window.location.origin,
      gateway_connected: true,
      control_plane_url: status.control_plane_url,
      control_plane_connected: Boolean(status.control_plane_connected),
      control_plane_error: status.control_plane_error,
      node_id: status.node_id,
      agent_count: status.agent_count,
      running_agent_count: status.running_agent_count,
    };
  } catch (exc) {
    return {
      gateway_url: window.location.origin,
      gateway_connected: false,
      gateway_error: String(exc),
      control_plane_connected: false,
    };
  }
}
